File: linkedlist/234_PalindromeLinkedList/234_PalindromeLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:        
        
        if head.next == None:
            return True

        first_pointer = head
        second_pointer = head       

        while first_pointer != None and second_pointer != None:                     

            if second_pointer.next == None:
                break
            else:                
                second_pointer = second_pointer.next.next

            first_pointer = first_pointer.next
        
        mid = first_pointer       
           
        if second_pointer != None:
            mid = mid.next        
           
        reverse = self.reverseLinkedList(mid)      

        while head != None and reverse != None:         
            if head.val != reverse.val:
                return False
            
            head = head.next
            reverse = reverse.next            
       
        return True
    
    def pushToTop(self, val, head: Optional[ListNode]) -> ListNode:
        node = ListNode(val)
        node.next = head
        
        return node

    def reverseLinkedList(self, head: Optional[ListNode]) -> ListNode:
        reverse = None

        while head != None:                        
            reverse = self.pushToTop(head.val, reverse)
            head = head.next

        return reverse

File: linkedlist/234_PalindromeLinkedList/234_PalindromeLinkedList.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	if head.Next == nil {
		return true
	}

	first_pointer := head
	second_pointer := head

	for second_pointer != nil {

		if second_pointer.Next != nil {
			second_pointer = second_pointer.Next.Next
		}

		if second_pointer == nil || second_pointer.Next == nil {
			break
		}

		first_pointer = first_pointer.Next
	}

	mid := first_pointer
	if second_pointer != nil {
		// even nodes
		mid = first_pointer.Next
	}

	//reverse rest half
	reverseList := new(ListNode)
	reverseList.Val = mid.Val

	//mid = mid.Next
	for mid.Next != nil {
		reverseList = pushToTop(mid.Next.Val, reverseList)
		mid = mid.Next
	}

	for reverseList != nil {

		if reverseList.Val != head.Val {
			return false
		}
		reverseList = reverseList.Next
		head = head.Next
	}

	return true
}

func pushToTop(val int, head *ListNode) *ListNode {

	node := new(ListNode)
	node.Val = val
	node.Next = head

	return node

}